snappy: add security generation and undo to install

mvo5 opened the pull request:
[based on https://github.com/ubuntu-core/snappy/pull/749]

Easy review via https://github.com/ubuntu-core/snappy/commit/a8c32186382c51bf40e09ae2fcb27b845b41452f

niemeyer commented:
LGTM

stevenwilkin commented:
:+1:

pedronis commented:
@mvo5 SetActive before this change would call the "activate" logic that would generate policies, with the extraction that is no longer the case, is that still correct? (no tests break, but are we missing tests? or generating policies is really orthogonal to activate/deactivate) ?

pedronis commented:
@mvo5 so after discussing it seems that here we need to break up "activate" and "deactivate" a just a little bit more (for example the optional deactivate at the beginning of activate should be its own thing, also because it needs to do the full thing)  and make sure SetActive still calls all the parts

mvo5 commented:
@pedronis moving the generation/removal to install/deinstall instead of activate/deactivate was expedient for the followup branch and because the names are unique per snap/develoepr/version(e.g. apparmor is named `hello-world.canonical_echo_6.0`) nothing should break. However it does not have to be done this way, we can also keep it in activate/deactivate and break those up furhter (this is what you seem to suggest in the following comment?).

pedronis commented on snappy/overlord.go:
```diff
@@ -447,6 +509,12 @@ func (o *Overlord) Uninstall(s *Snap, meter progress.Meter) error {
 	if active {
+		// deactivate current first
+		if current := ActiveSnapByName(s.Name()); current != nil {
+			if err := (&Overlord{}).SetActive(current, false, meter); err != nil {
```
this could just use DeactivateSnap directly?

pedronis commented on snappy/overlord.go:
```diff
@@ -246,6 +263,57 @@ func ActivateSnap(s *Snap, inter interacter) error {
+}
+
+// ActivateSnap is a wrapper around
+// (generate-security-profile, generate-wrappers, update-current-symlink)
```
a note about assuming nothing is currently active for this snap?

mvo5 replied:
I added a note.

pedronis commented:
LGTM with a couple comments, also I think maybe the code in UndoCopyDataSnap merits a comment about working in a state where nothing is active for the snap

niemeyer commented on snappy/overlord.go:
```diff
@@ -246,6 +263,57 @@ func ActivateSnap(s *Snap, inter interacter) error {
+// (generate-security-profile, generate-wrappers, update-current-symlink)
+func ActivateSnap(s *Snap, inter interacter) error {
+	currentActiveSymlink := filepath.Join(s.basedir, "..", "current")
+	currentActiveDir, _ := filepath.EvalSymlinks(currentActiveSymlink)
```
The err result here shouldn't be ignored. Below we take the lack of a currentActiveDir as a sign the snap is not activated, but it might have been an error.

niemeyer commented on snappy/overlord.go:
```diff
@@ -246,6 +263,57 @@ func ActivateSnap(s *Snap, inter interacter) error {
+		logger.Noticef("Failed to remove %q: %v", currentDataSymlink, err)
+	}
+
+	return nil
```
Why do we always return nil here even when it fails?

niemeyer commented on snappy/overlord.go:
```diff
@@ -246,6 +263,57 @@ func ActivateSnap(s *Snap, inter interacter) error {
+func removeCurrentSymlink(s *Snap, inter interacter) error {
+	currentActiveSymlink := filepath.Join(s.basedir, "..", "current")
+
+	// and finally the current symlink
```
The comment here needs to be updated or just removed.

niemeyer commented on snappy/overlord.go:
```diff
@@ -218,6 +208,33 @@ func ActivateSnap(s *Snap, inter interacter) error {
+}
+
+func removeGenerateWrappers(s *Snap, inter interacter) error {
+	// remove generated services, binaries, security policy
```
This might be moved up and updated to become the function comment.

niemeyer replied:
Also looks like the security policy isn't being removed here.

niemeyer commented on snappy/overlord.go:
```diff
@@ -218,6 +208,33 @@ func ActivateSnap(s *Snap, inter interacter) error {
+}
+
+func UndoGenerateWrappers(s *Snap, inter interacter) error {
+	return removeGenerateWrappers(s, inter)
```
Why do we have two local names for the same function?

niemeyer commented on snappy/overlord.go:
```diff
@@ -175,36 +175,26 @@ func UndoCopyData(newSnap *Snap, flags InstallFlags, meter progress.Meter) {
-	if err := generatePolicy(s.m, appsDir); err != nil {
-		return err
-	}
+	return removePolicy(s.m, s.basedir)
```
Why do we have two local names for the same function?

niemeyer commented on snappy/overlord.go:
```diff
@@ -175,36 +175,26 @@ func UndoCopyData(newSnap *Snap, flags InstallFlags, meter progress.Meter) {
-		}
-	}
+func UndoGenerateSecurityProfile(s *Snap) error {
+	return removeSecurityProfile(s)
```
Why do we have two local names for the same function?

niemeyer commented on snappy/overlord.go:
```diff
@@ -175,36 +175,26 @@ func UndoCopyData(newSnap *Snap, flags InstallFlags, meter progress.Meter) {
+	// Note that this must happen before binaries/services are
+	// generated because serices may get started
+	instDir := filepath.Join(dirs.SnapSnapsDir, s.Name(), s.Version())
+	return generatePolicy(s.m, instDir)
```
Why do we have (almost) two local names for the same function?  Can't we simply move the content of generatePolicy here and use that instead?

mvo5 commented:
@niemeyer @pedronis Thanks for the excellent review feedback. I addressed the review points now, please let me know if there is anything else I should fix.

pedronis commented:
+1

niemeyer commented on snappy/overlord.go:
```diff
@@ -246,6 +243,80 @@ func ActivateSnap(s *Snap, inter interacter) error {
+	} else if err1 != nil {
+		return fmt.Errorf("removeCurrentSymlink failed with: %v", err1)
+	} else if err2 != nil {
+		return fmt.Errorf("removeCurrentSymlink failed with: %v", err2)
```
Sorry for missing that earlier, but this is an error that may pop up to the user, so I'd suggest something like:

`cannot remove snap current symlink: ...`

niemeyer commented on snappy/overlord.go:
```diff
@@ -246,6 +243,80 @@ func ActivateSnap(s *Snap, inter interacter) error {
+
+	// there is already an active snap
+	if currentActiveDir != "" {
+		return fmt.Errorf("there is already an active snap: %v", currentActiveDir)
```
`cannot activate snap while another one is active: %v`

niemeyer commented on snappy/overlord.go:
```diff
@@ -274,21 +345,11 @@ func DeactivateSnap(s *Snap, inter interacter) error {
 		return err
```
Should the logic above be replaced by RemoveGeneratedWrappers?

Should it continue attempting to remove the follow up steps even if the first step fails? (solved if using the other func)

niemeyer commented on snappy/overlord.go:
```diff
@@ -218,6 +191,30 @@ func ActivateSnap(s *Snap, inter interacter) error {
+		logger.Noticef("Failed to remove desktop files for %q: %v", s.Name(), err)
+	}
+
+	return nil
```
It's strange that this returns an error that can never be set, even when it fails altogether.

niemeyer commented:
LGTM!
